File: synth/synthdata.py
import numpy as np
import math
import matplotlib.pyplot as plt
from scipy.interpolate import UnivariateSpline as UVS



def Model(X,ICF):
	L = len(ICF)
	MF=np.zeros((L),dtype='float')
	for i in range(0,L):
		MF[i]=np.sum(X*ICF[i,:])
	print 'Model: ',MF	
	return MF


def convolve(tdf_values,data_flux):
		val = len(data_flux)
		model_flux = np.empty(val)
		for idate in range(0,val):
			model_flux[idate] = np.sum(tdf_values*data_flux[idate])
		return model_flux

#TDF function from Blandford and McKee
def RAN(t,r):
	temp = np.empty([len(t)])
	for i in range(0,len(t)):
		if t[i] > 0 and t[i] <= (2.0*r):
			temp[i] = t[i]/(2.0*r)+np.log(2.0)
		elif t[i] > (2.0*r) and t[i] <= (4.0*r):
			temp[i] = 2.0 - t[i]/(2.0*r)+np.log(4.0*r/t[i])
		else:
			temp[i]=0
		#print 't: ', t[i], 'psi: ', temp[i]
	return 1.0/(2.0*r)*temp

'''IMPORT DATA FILES'''
#inputfile = '../Ver1/tdfL.txt'
#tdfL = np.loadtxt(open(inputfile,'rb'),delimiter=' ',skiprows=0)
#tdfL=tdfL[330]

inputfile = '../data/rvm_fluxes.csv'
rvm_spectra_flux = np.loadtxt(open(inputfile,'rb'),delimiter=',',skiprows=0)
#print 'Flux Array: ',rvm_spectra_flux.shape
spectra_scale=1e0
rvm_spectra_flux=rvm_spectra_flux*spectra_scale
#print rvm_spectra_flux.shape

inputfile = '../data/rvm_fluxes.csv'
data_flux = np.loadtxt(open(inputfile,'rb'),delimiter=',',skiprows=0)
#print('Flux Array: ',data_errflux.shape)
#nlines = len(data_errflux) # length first axis

inputfile = '../data/rvm_errfluxes.csv'
data_errflux = np.loadtxt(open(inputfile,'rb'),delimiter=',',skiprows=0)
#print('Error Flux Array: ',data_errflux.shape)
#nlines = len(data_errflux) # length first axis

inputfile = '../data/rvm_wavelengths.csv'
rvm_spectra_lam = np.loadtxt(open(inputfile,'rb'),delimiter=',',skiprows=0)
#print 'lam: ',len(rvm_spectra_lam)

inputfile = '../data/rvm_dates.csv'
rvm_spectra_date = np.loadtxt(open(inputfile,'rb'),delimiter=',',skiprows=0)
ndataperline = len(rvm_spectra_date)
#print 'date: ',len(rvm_spectra_date)

inputfile = '../data/arp151.b.dat'
data = np.loadtxt(open(inputfile,'rb'),delimiter=' ',skiprows=0)
continuum_date = data[:,0]
continuum_scale = 1.0e3
continuum_flux = data[:,1]*continuum_scale


inputfile = '../data/arp151.b.dat'
data = np.loadtxt(open(inputfile,'rb'),delimiter=' ',skiprows=0)
continuum_date = data[:,0]
continuum_scale = 1.0
continuum_flux = data[:,1]*continuum_scale
av = np.mean(continuum_flux)
print 'average: ', av
print continuum_flux
ax = np.max(continuum_flux)
continuum_errflux = data[:,2]

print('cont: ', data.shape)
'''DONE IMPORTING DATA FILES'''


'''----------------------------------------------------'''
'''===================================================='''
'''----------------------------------------------------'''
tau = 50

c = 1.0
r = 15.0
tau = np.linspace(0,50,49)
Psi = RAN(tau,r)	#TDF from Blanford & McKee
print tau
#plt.figure()
#plt.plot(tau,Psi)
#plt.show()



'''COMPUTING THE CONTINUUM FUNCTION FOR REQUIRED POINTS'''
pts_required=len(rvm_spectra_date)*len(tau)
interpolation_points = np.empty([len(rvm_spectra_date),len(tau)])
q=0
for idate in range(0,len(rvm_spectra_date)):
	for jdelay in range(0,len(tau)):
		interpolation_points[idate][jdelay] = rvm_spectra_date[idate]-tau[jdelay]
#f= interp1d(continuum_date,continuum_flux)					#Creates interpolation function
f= UVS(continuum_date,continuum_flux,k=2)					#Creates interpolation function

ICF = f(np.ravel(interpolation_points))			#Interpolates continuum over points
#print len(ICF)
#print len(rvm_spectra_date)*len(tau)
ICF=np.reshape(ICF,(len(rvm_spectra_date),len(tau)))
int_continuum_dates = interpolation_points					#only to make a matching set

ICF=ICF/1.0e6
print ICF.shape






#INTERPOLATION!!!
#f= interp1d(continuum_date,continuum_flux)  
#f = interp1d(data[0,:],data[1,:],bounds_error=False,fill_value=data[65,2])
#ICF = f(data_flux_dates)
#ICF[impulse_position]=impulse_value
print 'ICF: ', ICF.shape

L = Model(Psi,ICF)
LR=L
percent_err = 5.0
for j in range(0,len(data_errflux)):
	for h in range(0,len(L)):
		data_flux[j][h]=np.random.normal(L[h],0.02*L[h])
	#data_errflux[j] = percent_err/100*L

print data.shape

TDF = np.array([tau,Psi])
#print TDF
np.savetxt('rvm_flxS.csv',data_flux,delimiter=',')
print "Wrote: rvm_flxS.csv"
np.savetxt('rvm_flx_errS.csv',data_errflux,delimiter=',')	
print "Wrote: rvm_flx_errS.csv"
np.savetxt('continuumS.csv',data,delimiter=',')
print "Wrote: continuumS.csv"
np.savetxt('PsiS.txt',TDF,delimiter=',')

plt.figure(1)
#plt.plot(data_flux_dates,L)
plt.plot(data_flux[0])
plt.plot(L,'r--')
#plt.figure()
#plt.plot(data[:,1])
plt.show()


#print data[:,1]

File: synth/synthdata.jl
#using Optim
using PyPlot
#using Gadfly
include("synthlib.jl")
path="../data/"
#path ="synth/"
clear()



#=IMPORTING FILES=#
#SPECTRA
wavelength_filename = "rvm_wavelengths.csv"
wavelength_path = string(path,wavelength_filename)
wavelength = readcsv(wavelength_path)                    #List of measured wavelengths

spectra_filename= "rvm_fluxes.csv"
spectrapath = string(path,spectra_filename)
L = readcsv(spectrapath)                           #SPECTRAL FLUXES (L)
num_lines = size(L,1)                                #NUMBER OF SPECTRAL LINES

spectra_error_filename = "rvm_errfluxes.csv"
spectra_error_path = string(path,spectra_error_filename)
EL = readcsv(spectra_error_path)                 #SPECTRAL FLUX ERRORS

spectra_dates_filename = "rvm_dates.csv"
spectra_dates_path = string(path,spectra_dates_filename)
spectra_dates = readcsv(spectra_dates_path)                   #SPECTRAL SAMPLING DATES
num_spectra_samples = length(spectra_dates)      				  #NUMBER OF DATA POINTS IN SPECTRA

#CONTINUUM
continuum_array_filename = "arp151.b.dat"
continuum_array_path =string(path,continuum_array_filename)
continuum_array = readdlm(continuum_array_path)
#continuum_array = readcsv(continuum_array_path)
println(size(continuum_array))
#CONTINUUM_ARRAY CONTAINTS THE CONTINUUM DATES, FLUX, AND FLUX ERRORS.
#IN THAT ORDER.

continuum_dates = continuum_array[:,1]
continuum_scale = 1.0
continuum_flux = continuum_array[:,2]*continuum_scale
continuum_error_flux = continuum_array[:,3]*continuum_scale
num_continuum_dates = length(continuum_dates)


########################
#=DONE IMPORTING FILES=#
########################


#=SETTING UP TIME DELAY FUNCTION=#

num_tdf_times = 50
tdf_values = zeros(num_tdf_times)
#tdf_times = linspace(1,20,num_tdf_times)
tdf_times = collect(1.0:(19.0/num_tdf_times):20.0)
X =   zeros((num_tdf_times))
initial_x = 0.04																#INITIAL VALUE FOR THE TDF
fill!(X,initial_x)															#FILL TDF WITH INITIAL VALUE
#=DONE SETTING UP TIME DELAY FUNCTION=#

#= Creating Artificial Data Arrays =#
num_lines = 100
num_spectra_samples = 100

#=SET UP VELOCITY DELAY MAP=#
num_tdf_times = 50
VDM=zeros((num_tdf_times,num_lines))


#= CST data =#
fill!(VDM, initial_x)

#= 1D data =#



#=2D data =#
#for i in 1:num_tdf_times
#	for j in 1:num_lines
#		VDM[i,j] = sin( (sin(j/100.0*pi)*i)/50.0*pi)
#	end
#end

println(VDM)
#=DONE SETTING UP VDM=#



#For Wavelength:
low = 450.0
high = 1000.0
Lrange = high - low
wavelengths = Lrange * sort(rand(num_lines))+low


#For Spectra Dates/Samples:
beginning = minimum(spectra_dates)
ending = maximum(spectra_dates)
range  = ending -beginning
spectra_dates = range * sort(rand(num_spectra_samples))+beginning




#= COMPUTING THE CONTINUUM FUNCTION FOR REQUIRED POINTS =#

interpolation_points = zeros(num_spectra_samples,num_tdf_times)
H = zeros(num_spectra_samples,num_tdf_times)
HE= zeros(num_spectra_samples,num_tdf_times)
for date in collect(1:num_spectra_samples)
	for delay in collect(1:num_tdf_times)
		interpolation_points[date,delay]=spectra_dates[date]-tdf_times[delay]
  end
  P = collect(interpolation_points[date,:])
  H[date,:] =interp(interpolation_points[date,:],continuum_dates,continuum_flux)
  HE[date,:] = interp(interpolation_points[date,:],continuum_dates,continuum_error_flux)
end



#println("Size of L: ", size(L))

synthetic_data = Model(VDM,H)		

#= Now have the synthetic spectral line (L) and the TDF (x).
Now need to build the output arrays to match the arp151 data. =#


#OUTPUT ARRAYS
#= SETTING UP OUTPUT ARRAYS =#

println("NUM LINES: ",num_lines)
println("NUM SPECTRA: ", num_spectra_samples)
flx_arr = zeros(num_lines,num_spectra_samples)
err_arr = zeros(num_lines,num_spectra_samples)
TDF = zeros(num_tdf_times)

println("Data Array: ", size(flx_arr))



#FILL OUPUT ARRAYS!
sigma = 3.0															#FOR ADDING ERROR
for j in 1:num_lines
	for h in 1:num_spectra_samples
		flx_arr[j,h] = synthetic_data[h]+sigma*randn(1)[1]		
		err_arr[j,h] = sigma
		#flx_arr[j,:] = synd										#NO ERROR
 		#err_arr[j,h] = 1.0											#NO ERROR 
	end
end

#WRITE OUPUT FILES

println("--------------------------")
println(flx_arr[1,:])
println("--------------------------")

#writecsv("filename.csv",array)
writecsv("wavelengthS.csv",wavelengths)
writecsv("rvm_dateS.csv",spectra_dates)
writecsv("rvm_flxS.csv", flx_arr)
writecsv("rvm_flx_errS.csv",err_arr)
writecsv("continuumS.csv",continuum_array)
writecsv("TDF.csv",VDM)
#println(size(flx_arr),size(err_arr),size(continuum_array),size(TDF))
#println(L[1,:])

figure(1)
#plot(synthetic_data,"b") #PLOT MODEL
#plot(flx_arr[1,:]',"r*")		 #PLOT snyth noise
imshow(VDM)
colorbar()
show()


println("Done")
